Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -252,6 +263,16 @@ public class HTTPClient {
                                                               delegate: Delegate,
                                                               eventLoop eventLoopPreference: EventLoopPreference,
                                                               deadline: NIODeadline? = nil) -> Task<Delegate.Response> {
+        // FIXME: Is this correct to assert?
+        self.stateLock.withLock {
+            switch state {
+            case .upAndRunning:
+                break
+            case .shuttingDown, .shutDown:
+                assertionFailure("Attempting to execute a request on shutdown client")
+            }
+        }
```

Comment:
@weissi Do you think this is correct to assert here?

Reply:
@adtrevor I'd probably just fail the request here too.

Reply:
@weissi If state is `.shutDown` and `EventLoopGroupProvider` is `createNew`, it means the client `EventLoopGroup` will be shutdown, so if we return a failed future I'm worried code waiting on this result will just wait forever as the completion callbacks will never be called?

Reply:
@adtrevor you've actually got a good point. However, it's not your responsibility to check this in AsyncHTTPClient. There are _loads_ of places that would start hanging if someone just randomly shut down the EventLoop(Group).

Chatting about this with @Lukasa made us realise that NIO should handle this case in a better way and I filed https://github.com/apple/swift-nio/issues/1383 . So I think you should just fail the promise and defer to NIO to do the right thing.

Reply:
@weissi 
> There are loads of places that would start hanging if someone just randomly shut down the EventLoop(Group)
Happened a few times to me indeed, nice to see it will be fixed in NIO 👍

Reply:
@weissi Fixed in https://github.com/swift-server/async-http-client/pull/105/commits/828c3f4c07a324eba0b74311f7444d3a023589cf